code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Optimize.java

```diff
@@ -272,4 +277,191 @@ private static void failWithError(final AdobeCallback<?> callback, final AdobeEr
+                                propositionsMap.put(OptimizeUtils.retrieveSurfacePathFromScope(proposition.getScope()), proposition);
+                            }
+                        }
+                    }
+                    callback.call(propositionsMap);
+                } catch (DataReaderException e) {
+                    failWithError(callback, AdobeError.UNEXPECTED_ERROR);
+                }
+            }
+        });
+    }
+
+    /**
+     * This API registers a permanent callback which is invoked whenever the Edge extension dispatches an Event handle,
+     * upon a personalization decisions response from the Experience Edge Network.
+     * <p>
+     * The personalization query requests can be triggered by the {@link Optimize#updatePropositionsForSurfacePaths(List, Map, Map)} API.
+     *
+     * @param callback {@code AdobeCallbackWithError<Map<String, Proposition>>} which will be invoked when decision propositions are received from the Edge network.
+     */
+    public static void setPropositionsHandler(@NonNull final AdobeCallback<Map<String, Proposition>> callback) {
+        if(!isPropositionsListenerRegistered) {
+            MobileCore.registerEventListener(OptimizeConstants.EventType.OPTIMIZE, OptimizeConstants.EventSource.NOTIFICATION, new AdobeCallbackWithError<Event>() {
+                @Override
+                public void fail(final AdobeError error) {
+                }
+
+                @Override
+                public void call(final Event event) {
+                    final Map<String, Object> eventData = event.getEventData();
+                    if (OptimizeUtils.isNullOrEmpty(eventData)) {
+                        return;
+                    }
+
+                    final List<Map<String, Object>> propositionsList;
+                    try {
+                        propositionsList = DataReader.getTypedListOfMap(Object.class, eventData, OptimizeConstants.EventDataKeys.PROPOSITIONS);
+
+                        final Map<String, Proposition> propositionsMap = new HashMap<>();
+                        if (!OptimizeUtils.isNullOrEmpty(propositionsList)) {
+                            for (final Map<String, Object> propositionData : propositionsList) {
+                                final Proposition proposition = Proposition.fromEventData(propositionData);
+                                if (proposition != null && !OptimizeUtils.isNullOrEmpty(proposition.getScope())) {
+                                    propositionsMap.put(OptimizeUtils.retrieveSurfacePathFromScope(proposition.getScope()), proposition);
+                                }
+                            }
+                        }
+
+                        if (!propositionsMap.isEmpty()) {
+                            callback.call(propositionsMap);
+                        }
+                    } catch (DataReaderException ignored) {
```

Comment:
`final` DataReaderException e. Probably good to add a debug log here than just ignore exception.

Reply:
This case should never occur since the Optimize SDK creates and dispatches the event